src/bitcoin_node_cli.cpp

```diff
@@ -1,60 +1,86 @@
+		}
+	});
+	m_ptr = curl_easy_init();
+	if (m_ptr == nullptr) {
+		throw std::runtime_error("CURL init error");
+	}
+}
+
+c_curl_ptr::~c_curl_ptr() {
+	curl_easy_cleanup(m_ptr);
+	curl_global_cleanup();
+}
+
+CURL *c_curl_ptr::get_raw_ptr() const {
+	return m_ptr;
+}
+
+//////////////////////////////////////////////////////
+
 bitcoin_node_cli::bitcoin_node_cli(const std::string &ip_address, unsigned short port)
 :
-	m_io_service(),
-	m_btc_rpc_endpoint(boost::asio::ip::address_v4::from_string(ip_address), port)
+	m_rpc_http_address("http://" + ip_address + ':' + std::to_string(port))
 {
 }
 
-uint32_t bitcoin_node_cli::get_balance() {
-	try {
-		// TODO: bitcoin rpc pass
-		std::string get_balance_request;
-		get_balance_request += "POST / HTTP/1.1\r\n";
-		get_balance_request += "Host: ::1\r\n";
-		get_balance_request += "Connection: close\r\n";
-		get_balance_request += "Authorization: Basic X19jb29raWVfXzowMmNmYzBlYjE2YWU1OTQzMDIxMWVlZTc0NTc0NGVhNTVkZTk4MjM2ZDkzNTU2YzE5MGMxNzEyNjgyNTQwNmNh\r\n";
-		get_balance_request += "Content-Length: 48\r\n\r\n";
-		get_balance_request += R"({"method":"getbalance","params":["*",0],"id":1})";
-		get_balance_request += "\n";
-
-		boost::asio::ip::tcp::socket socket(m_io_service);
-		socket.connect(m_btc_rpc_endpoint);
-		socket.send(boost::asio::buffer(get_balance_request));
-		std::string receive_buffer(1024, '\0');
-		size_t readed_bytes = socket.receive(boost::asio::buffer(&receive_buffer[0], receive_buffer.size()));
-
-		receive_buffer.resize(readed_bytes);
-		auto it = std::find(receive_buffer.begin(), receive_buffer.end(), '{'); // find begin of json data
-		receive_buffer.erase(receive_buffer.begin(), it); // remove http POST data
-		_mark(receive_buffer);
-
-		// remove other json fields
-		receive_buffer.erase(receive_buffer.begin()); // remove first '{'
-		auto pos = receive_buffer.find(R"("result":)"); // find "result":
-		if (pos == std::string::npos) return 0;
-		receive_buffer.erase(pos, std::string(R"("result":)").size()); // remove "result":
-		it = std::find_if(receive_buffer.begin(), receive_buffer.end(),
-		                  [](unsigned char c) {
-		                  		if(c == '.' || std::isdigit(c)) return false;
-		                  		return true;
-		                  }
-		);
-		receive_buffer.erase(it, receive_buffer.end());
-		const std::string btc_amount_str = receive_buffer;
-
-		_mark("btc_amount_str " << btc_amount_str);
-		std::stringstream ss;
-		ss << std::setprecision(20) << btc_amount_str;
-		double btc_amount;
-		ss >> btc_amount;
-
-		return btc_amount * 10'000'000. ; // return balance in satoshi
-	} catch (const std::exception &e) {
-		_erro(e.what());
-		abort();
-		return 0;
+uint32_t bitcoin_node_cli::get_balance() const {
+	const std::string request (R"({"method":"getbalance","params":["*",0],"id":1})");
+	const std::string receive_data = send_request_and_get_response(request);
+	_mark("Receive data " << receive_data);
+
+	std::setlocale(LC_ALL, "en_US.UTF-8");
```

Comment:
we can not do this - possible UB (depending on what other code is doing)

"
Because setlocale modifies global state which affects execution of locale-dependent functions, it is undefined behavior to call it from one thread, while another thread is executing any of the following functions
"

this work around for JSON library bug should be improved